src/coreclr/src/vm/jitinterface.cpp

```diff
@@ -4477,7 +4477,7 @@ TypeCompareState CEEInfo::compareTypesForCast(
 #endif // FEATURE_COMINTEROP
 
     // If casting from ICastable, don't try to optimize
-    if (fromHnd.GetMethodTable()->IsICastable())
+    if (fromHnd.GetMethodTable()->IsICastable() || fromHnd.GetMethodTable()->IsICastableObject())
```

Comment:
A matching change should be also done in https://github.com/dotnet/runtime/blob/b934bc02abe2bf03e8efd027a2d3bc14ba011acf/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs#L1773

Reply:
Do we have a tests that would catch this bug?

Reply:
Could you point me at what the best way to get at this information from there? I know just iterating over the `RuntimeInterfaces` on the `TypeDesc` and matching the name and namespace to `System.Runtime.CompilerServices.ICastableObject` works, but I don't know what a good way to do this would be.

The tests I added would go down this path when crossgen2 is run on them. However, if `compareTypesForCast` ends up with `MustNot` from the normal logic, we switch it to `May` to allow for a previously negative result to become positive, so they would not actually catch this bug: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs#L1837-L1846

Reply:
@MichalStrehovsky Could you please recommend the best way to add ICastableObject support to the managed type system?

Reply:
Since this will be public API that concerns the type system, we can probably make this a first class citizen in the type system. I would:

* Add new flags `IsCastableObject` and `IsCastableObjectComputed` to the `TypeFlags` enum.
* Add `IsCastableObject` property to `TypeDesc`. The implementation for that will look very similar to the existing `TypeDesc.HasStaticConstructor`
* In `TypeSystemContext.ComputeTypeFlags` add handling for the new flag (we probably just want to walk the `RuntimeInterfaces` and for each interface, call a new abstract method on `TypeSystemContext (e.g. `IsICastableObjectInterface`).
* Add the abstract `bool TypeSystemContext.IsICastableObjectInterface(DefType type)` to `TypeSystemContext`.
* Implement the abstract `IsICastableObjectInterface` in the `MetadataTypeSystemContext` class. The implementation will basically just `MetadataType t = (MetadataType)type; return t.Module == SystemModule && t.Name == "ICastableObject" t.Namespace == "...")` (the cast from `DefType` to `MetadataType` is always safe in the `Metadata*.cs` files).

Reply:
Thanks @MichalStrehovsky! That is super helpful.